--- collectd-4.5.0/src/interface.c.peak	2008-09-17 09:01:06.000000000 +0200
+++ collectd-4.5.0/src/interface.c	2008-09-17 09:07:47.000000000 +0200
@@ -261,7 +261,7 @@ static int interface_read (void)
 #elif KERNEL_LINUX
 	FILE *fh;
 	char buffer[1024];
-	unsigned long long incoming, outgoing;
+	unsigned long long incoming, outgoing, in_packets, out_packets;
 	char *device;
 	
 	char *dummy;
@@ -295,13 +295,15 @@ static int interface_read (void)
 		if (numfields < 11)
 			continue;
 
+		in_packets = atoll (fields[1]);
+		out_packets = atoll (fields[9]);
+
 		incoming = atoll (fields[0]);
 		outgoing = atoll (fields[8]);
+		if ( (out_packets < 500) && (in_packets < 500) )  continue;
 		if_submit (device, "if_octets", incoming, outgoing);
 
-		incoming = atoll (fields[1]);
-		outgoing = atoll (fields[9]);
-		if_submit (device, "if_packets", incoming, outgoing);
+		if_submit (device, "if_packets", in_packets, out_packets);
 
 		incoming = atoll (fields[2]);
 		outgoing = atoll (fields[10]);
