diff -up linux-2.6.26.2/drivers/rtc/rtc-x1205.c.pico_false_rtc linux-2.6.26.2/drivers/rtc/rtc-x1205.c
--- linux-2.6.26.2/drivers/rtc/rtc-x1205.c.pico_false_rtc	2009-03-25 15:31:14.000000000 +0100
+++ linux-2.6.26.2/drivers/rtc/rtc-x1205.c	2009-03-25 15:35:23.000000000 +0100
@@ -182,6 +182,11 @@ static int x1205_set_datetime(struct i2c
 
 	/* should we also set the date? */
 	if (datetoo) {
+		if ( tm->tm_year < 195 ){
+			//dev_info(&client->dev,"year is zerwixt, got %d\n",tm->tm_year);
+			tm->tm_year+=1900;
+		}
+
 		dev_dbg(&client->dev,
 			"%s: mday=%d, mon=%d, year=%d, wday=%d\n",
 			__func__,
