--- src/appl/bsd/krlogin.c.errno	2006-09-14 09:53:27.000000000 +0200
+++ src/appl/bsd/krlogin.c	2006-09-14 09:53:45.000000000 +0200
@@ -69,7 +69,7 @@
 #include <sys/types.h>
 #include <sys/param.h>
 #include <sys/ioctl.h>
-#include <sys/errno.h>
+#include <errno.h>
 #include <sys/file.h>
 #include <sys/socket.h>
 #include <sys/time.h>
--- src/lib/rpc/bindresvport.c.errno	2006-09-14 09:54:02.000000000 +0200
+++ src/lib/rpc/bindresvport.c	2006-09-14 09:54:13.000000000 +0200
@@ -37,7 +37,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <sys/types.h>
-#include <sys/errno.h>
+#include <errno.h>
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <gssrpc/rpc.h>
--- src/lib/rpc/clnt_generic.c.errno	2006-09-14 09:52:34.000000000 +0200
+++ src/lib/rpc/clnt_generic.c	2006-09-14 09:52:47.000000000 +0200
@@ -36,7 +36,7 @@
 #include <string.h>
 #include <gssrpc/rpc.h>
 #include <sys/socket.h>
-#include <sys/errno.h>
+#include <errno.h>
 #include <netdb.h>
 #include "autoconf.h"
 
--- src/lib/rpc/auth_gssapi.c.errno	2006-09-14 09:51:45.000000000 +0200
+++ src/lib/rpc/auth_gssapi.c	2006-09-14 09:51:54.000000000 +0200
@@ -5,7 +5,7 @@
 
 #include <stdio.h>
 #include <string.h>
-#include <sys/errno.h>
+#include <errno.h>
 
 #include <gssapi/gssapi.h>
 #include <gssapi/gssapi_generic.h>
