--- aescrypt-2.1/aescrypt.c.rindex	2008-08-24 20:35:21.000000000 +0200
+++ aescrypt-2.1/aescrypt.c	2008-08-24 20:37:08.000000000 +0200
@@ -692,7 +692,7 @@ void usage(const char *progname)
 {
     const char* progname_real; //contains the real name of the program (without path)
 
-    progname_real = rindex(progname, '/');
+    progname_real = strrchr(progname, '/');
     if (progname_real == NULL) //no path in progname: use progname
     {
         progname_real = progname;
@@ -715,7 +715,7 @@ void version(const char *progname)
 {
     const char* progname_real; //contains the real name of the program (without path)
 
-    progname_real = rindex(progname, '/');
+    progname_real = strrchr(progname, '/');
     if (progname_real == NULL) //no path in progname: use progname
     {
         progname_real = progname;
