--- openswan2-2.6.04/programs/pluto/ikev1_quick.c.check_keymat	2008-01-25 19:38:47.000000000 +0100
+++ openswan2-2.6.04/programs/pluto/ikev1_quick.c	2008-01-31 23:18:53.000000000 +0100
@@ -255,7 +255,7 @@
 		break;
 	    default:
 #ifdef KERNEL_ALG
-	      if (kernel_alg_esp_auth_ok(pi->attrs.transattrs.integ_hash, NULL)) {
+	      if (kernel_alg_esp_auth_ok(pi->attrs.transattrs.integ_hash, NULL) == NULL) {
 		  needed_len += kernel_alg_esp_auth_keylen(pi->attrs.transattrs.integ_hash);
 		  break;
 	      } 
