--- dietlibc-0.30/include/sys/inotify.h.inotify	2005-10-04 19:47:03.000000000 +0200
+++ dietlibc-0.30/include/sys/inotify.h	2007-12-14 20:29:16.000000000 +0100
@@ -35,6 +35,9 @@
 #define IN_MOVE			(IN_MOVED_FROM | IN_MOVED_TO) /* moves */
 
 /* special flags */
+#define IN_ONLYDIR		0x01000000	/* only watch the path if it is a directory */
+#define IN_DONT_FOLLOW		0x02000000	/* don't follow a sym link */
+#define IN_MASK_ADD		0x20000000	/* add to the mask of an already existing watch */
 #define IN_ISDIR		0x40000000	/* event occurred against dir */
 #define IN_ONESHOT		0x80000000	/* only send event once */
 
