--- dietlibc-0.33/lib/ttyname.c.ttyname_r	2010-09-03 23:12:39.000000000 +0200
+++ dietlibc-0.33/lib/ttyname.c	2010-09-03 23:56:29.000000000 +0200
@@ -7,19 +7,17 @@
 
 #include <stdlib.h>
 
-char *ttyname(int fd) {
+char *ttyname_r(int fd, char *buf, size_t buflen) {
 #ifdef SLASH_PROC_OK
   char ibuf[20];
-  static char obuf[20];
   int len;
   if (!isatty(fd)) return 0;
   strcpy(ibuf,"/proc/self/fd/");
   ibuf[__ltostr(ibuf+14,6,(unsigned long)fd,10,0)+14]=0;
-  if ((len=readlink(ibuf,obuf,sizeof(obuf)-1))<0) return 0;
-  obuf[len]=0;
-  return obuf;
+  if ((len=readlink(ibuf,buf,buflen-1))<0) return 0;
+  buf[len]=0;
+  return buf;
 #else
-  static char buf[20];
   struct stat s;
   char *c=buf+8;
   int n;
@@ -63,4 +61,9 @@ duh:
 #endif
 }
 
+char *ttyname(int fd) {
+  char buf[20];
+  return ttyname_r(fd, buf, sizeof(buf));
+}
+
 #endif
