--- dietlibc-0.30/libdl/_dl_relocate.c.libdl	2005-05-13 14:14:06.000000000 +0200
+++ dietlibc-0.30/libdl/_dl_relocate.c	2006-11-04 12:40:56.000000000 +0100
@@ -46,7 +46,14 @@
 
 #ifdef __i386__
   if (typ==R_386_32) {			/* 1 */
-    *loc=(unsigned long)(dh->mem_base+dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+    unsigned long value=(unsigned long)(dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+#ifdef DEBUG
+    pf(__func__); pf(" value "); ph(value); pf("\n");
+#endif
+    if (value)
+      *loc=(unsigned long)(dh->mem_base+value);
+    else
+      *loc=(unsigned long)_dl_sym(dh,ELF_R_SYM(rel->r_info));
   } else if (typ==R_386_COPY)  {	/* 5 */
     unsigned long len=dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_size;
     void*from=_dl_sym_next(dh,ELF_R_SYM(rel->r_info));
@@ -66,7 +73,11 @@
 #endif
 #ifdef __arm__
   if (typ==R_ARM_ABS32) {		/*  2 */
-    *loc=(unsigned long)(dh->mem_base+dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+    unsigned long value=(unsigned long)(dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+    if (value)
+      *loc=(unsigned long)(dh->mem_base+value);
+    else
+      *loc=(unsigned long)_dl_sym(dh,ELF_R_SYM(rel->r_info));
   } else if (typ==R_ARM_COPY)  {	/* 20 */
     unsigned long len=dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_size;
     void*from=_dl_sym_next(dh,ELF_R_SYM(rel->r_info));
@@ -86,7 +97,11 @@
 #endif
 #ifdef __x86_64__
   if (typ==R_X86_64_64) {			/* 1 */
-    *loc=(unsigned long)(dh->mem_base+dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+    unsigned long value=(unsigned long)(dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_value);
+    if (value)
+      *loc=(unsigned long)(dh->mem_base+value);
+    else
+      *loc=(unsigned long)_dl_sym(dh,ELF_R_SYM(rel->r_info));
   } else if (typ==R_X86_64_COPY)  {	/* 5 */
     unsigned long len=dh->dyn_sym_tab[ELF_R_SYM(rel->r_info)].st_size;
     void*from=_dl_sym_next(dh,ELF_R_SYM(rel->r_info));
