--- dietlibc-0.30/libpthread/pthread_fread.c.threadsafe_fread	2009-04-22 21:34:42.000000000 +0200
+++ dietlibc-0.30/libpthread/pthread_fread.c	2009-04-22 21:34:26.000000000 +0200
@@ -0,0 +1,9 @@
+#include <dietstdio.h>
+
+size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+  size_t tmp;
+  pthread_mutex_lock(&stream->m);
+  tmp=fread_unlocked(ptr,size,nmemb,stream);
+  pthread_mutex_unlock(&stream->m);
+  return tmp;
+}
