--- dietlibc-0.30/lib/__v_printf.c.intmax	2005-04-10 15:59:32.000000000 +0200
+++ dietlibc-0.30/lib/__v_printf.c	2006-08-07 16:35:20.000000000 +0200
@@ -81,9 +81,15 @@
       case 'h':
 	--flag_long;
 	goto inn_printf;
+#if __WORDSIZE != 64
+      case 'j':
+#endif
       case 'q':		/* BSD ... */
       case 'L':
 	++flag_long; /* fall through */
+#if __WORDSIZE == 64
+      case 'j':
+#endif
       case 'l':
 	++flag_long;
 	goto inn_printf;
 
