--- dietlibc-0.30/arm/dyn_syscalls.S.select	2008-07-01 18:05:01.000000000 +0200
+++ dietlibc-0.30/arm/dyn_syscalls.S	2008-07-01 18:03:49.000000000 +0200
@@ -37,6 +37,7 @@ __unified_syscall:
 /* ok now include all syscalls.s (*.S) and sysdep *.S */
 #include "mmap.S"
 #include "waitpid.S"
+#include "select.S"
 
 #include "../syscalls.s/__getpagesize.S"
 #include "../syscalls.s/__pread.S"
@@ -142,7 +143,6 @@ __unified_syscall:
 #include "../syscalls.s/sched_setparam.S"
 #include "../syscalls.s/sched_setscheduler.S"
 #include "../syscalls.s/sched_yield.S"
-#include "../syscalls.s/select.S"
 #include "../syscalls.s/sendfile.S"
 #include "../syscalls.s/sendfile64.S"
 #include "../syscalls.s/setdomainname.S"
