--- dietlibc-0.30/include/sys/inotify.h.struct_inotify_event	2010-09-07 00:00:48.000000000 +0200
+++ dietlibc-0.30/include/sys/inotify.h	2010-09-07 21:08:56.000000000 +0200
@@ -1,14 +1,14 @@
 #ifndef _SYS_INOTIFY_H
 #define _SYS_INOTIFY_H
 
-#include <sys/cdefs.h>
+#include <stdint.h>
 
 struct inotify_event {
-	__s32		wd;		/* watch descriptor */
-	__u32		mask;		/* watch mask */
-	__u32		cookie;		/* cookie to synchronize two events */
-	__u32		len;		/* length (including nulls) of name */
-	char		name[0];	/* stub for possible name */
+	int  		wd;		/* watch descriptor */
+	uint32_t	mask;		/* watch mask */
+	uint32_t	cookie;		/* cookie to synchronize two events */
+	uint32_t	len;		/* length (including nulls) of name */
+	char		name[];		/* stub for possible name */
 };
 
 /* the following are legal, implemented events that user-space can watch for */
