--- dietlibc-cvs/Makefile.misc	2005-10-04 19:47:03.000000000 +0200
+++ dietlibc-cvs/Makefile	2005-11-26 22:02:19.000000000 +0100
@@ -86,11 +86,10 @@
 
 WHAT=	$(OBJDIR) $(OBJDIR)/start.o $(OBJDIR)/dyn_start.o $(OBJDIR)/dyn_stop.o \
 	$(OBJDIR)/dietlibc.a $(OBJDIR)/liblatin1.a \
-	$(OBJDIR)/libcompat.a $(OBJDIR)/libm.a \
+	$(OBJDIR)/libcompat.a $(OBJDIR)/libm.a $(OBJDIR)/libdl.a \
 	$(OBJDIR)/librpc.a $(OBJDIR)/libpthread.a \
 	$(OBJDIR)/libcrypt.a \
-	$(OBJDIR)/diet $(OBJDIR)/diet-i $(OBJDIR)/elftrunc \
-	$(OBJDIR)/dnsd
+	$(OBJDIR)/diet $(OBJDIR)/diet-i
 
 all: $(WHAT)
 
@@ -137,6 +136,7 @@
 COMMENT = :
 endif
 CFLAGS += -Wall -W -Wchar-subscripts -Wmissing-prototypes -Wmissing-declarations -Wno-switch -Wno-unused -Wredundant-decls
+CFLAGS += $(MORE_CFLAGS) -D__linux__
 
 PWD=$(shell pwd)
 
@@ -269,6 +269,8 @@
 
 DYN_LIBMATH_OBJS = $(patsubst $(OBJDIR)/%.o,$(PICODIR)/%.o,$(LIBMATHOBJ))
 
+DYN_LIBMATH_OBJS = $(patsubst $(OBJDIR)/%.o,$(PICODIR)/%.o,$(LIBMATHOBJ))
+
 $(PICODIR)/libc.so: $(PICODIR) $(DYN_LIBC_OBJ)
 	$(LD_UNSET) $(CROSS)$(CC) -nostdlib -shared -o $@ $(CFLAGS) -fPIC $(DYN_LIBC_OBJ) -lgcc -Wl,-soname=libc.so
 
@@ -288,6 +290,9 @@
 $(PICODIR)/libm.so: $(DYN_LIBMATH_OBJS) dietfeatures.h
 	$(LD_UNSET) $(CROSS)$(CC) -nostdlib -shared -o $@ $(CFLAGS) -fPIC $(DYN_LIBMATH_OBJS) -L$(PICODIR) -lc -Wl,-soname=libm.so
 
+#$(PICODIR)/libm.so: $(DYN_LIBMATH_OBJS) dietfeatures.h
+#	$(CROSS)$(CC) -nostdlib -shared -o $@ $(CFLAGS) -fPIC $(DYN_LIBMATH_OBJS) -L$(PICODIR) -lc -Wl,-soname=libm.so
+#
 
 $(SYSCALLOBJ): syscalls.h
 
@@ -349,17 +354,16 @@
 t1:
 	$(CROSS)$(CC) -g -o t1 t.c
 
-install: $(OBJDIR)/start.o $(OBJDIR)/dietlibc.a $(OBJDIR)/librpc.a $(OBJDIR)/liblatin1.a $(OBJDIR)/libcompat.a $(OBJDIR)/elftrunc $(OBJDIR)/diet-i
+install: $(OBJDIR)/start.o $(OBJDIR)/dietlibc.a $(OBJDIR)/librpc.a $(OBJDIR)/liblatin1.a $(OBJDIR)/libcompat.a $(OBJDIR)/diet-i
 	$(INSTALL) -d $(DESTDIR)$(ILIBDIR) $(DESTDIR)$(MAN1DIR) $(DESTDIR)$(BINDIR)
 	$(INSTALL) $(OBJDIR)/start.o $(DESTDIR)$(ILIBDIR)/start.o
-	$(INSTALL) -m 644 $(OBJDIR)/libm.a $(OBJDIR)/libpthread.a $(OBJDIR)/librpc.a \
-$(OBJDIR)/liblatin1.a $(OBJDIR)/libcompat.a $(OBJDIR)/libcrypt.a $(DESTDIR)$(ILIBDIR)
+	$(INSTALL) -m 644 $(OBJDIR)/libm.a $(OBJDIR)/libpthread.a $(OBJDIR)/libdl.a $(OBJDIR)/librpc.a \
+	$(OBJDIR)/liblatin1.a $(OBJDIR)/libcompat.a $(OBJDIR)/libcrypt.a $(DESTDIR)$(ILIBDIR)
 	$(INSTALL) -m 644 $(OBJDIR)/dietlibc.a $(DESTDIR)$(ILIBDIR)/libc.a
 ifeq ($(MYARCH),$(ARCH))
 	$(INSTALL) $(OBJDIR)/diet-i $(DESTDIR)$(BINDIR)/diet
 	-$(INSTALL) $(PICODIR)/diet-dyn-i $(DESTDIR)$(BINDIR)/diet-dyn
 endif
-	$(INSTALL) -m 755 $(OBJDIR)/elftrunc $(OBJDIR)/dnsd $(DESTDIR)$(BINDIR)
 	-$(INSTALL) $(OBJDIR)/pstart.o $(OBJDIR)/libgmon.a $(OBJDIR)/dyn_start.o $(OBJDIR)/dyn_stop.o $(DESTDIR)$(ILIBDIR)
 	-$(INSTALL) $(PICODIR)/libc.so $(DESTDIR)$(ILIBDIR)/libc.so
 	-$(INSTALL) $(PICODIR)/libpthread.so $(DESTDIR)$(ILIBDIR)/libpthread.so
--- dietlibc-cvs/lib/alloc.c.misc	2005-03-15 09:51:22.000000000 +0100
+++ dietlibc-cvs/lib/alloc.c	2005-11-26 21:48:05.000000000 +0100
@@ -78,8 +78,6 @@
 }
 
 /* small mem */
-static void __small_free(void*_ptr,size_t _size) REGPARM(2);
-
 static void REGPARM(2) __small_free(void*_ptr,size_t _size) {
   __alloc_t* ptr=BLOCK_START(_ptr);
   size_t size=_size;
--- dietlibc-cvs/lib/__lltostr.c.misc	2001-08-14 18:56:49.000000000 +0200
+++ dietlibc-cvs/lib/__lltostr.c	2005-11-26 21:48:05.000000000 +0100
@@ -6,6 +6,7 @@
 {
   char *tmp;
   unsigned int j=0;
+  char A=(UpCase?'A':'a')-'9'-1;
 
   s[--size]=0;
 
@@ -23,7 +24,7 @@
   while((tmp>s)&&(i))
   {
     tmp--;
-    if ((*tmp=i%base+'0')>'9') *tmp+=(UpCase?'A':'a')-'9'-1;
+    if ((*tmp=i%base+'0')>'9') *tmp+=A;
     i=i/base;
     j++;
   }
