--- e2fsprogs-1.32/misc/e2label.c.printf	2003-12-12 10:33:27.000000000 +0100
+++ e2fsprogs-1.32/misc/e2label.c	2003-12-12 10:40:36.000000000 +0100
@@ -54,21 +54,23 @@
 	fd = open(dev, mode);
 	if (fd < 0) {
 	     perror(dev);
-	     fprintf (stderr, _("e2label: cannot open %s\n"), dev);
+	     write(2,"e2label: cannot open ", 21);
+	     write(2, dev, strlen(dev));
+	     write(2, "\n", 1);
 	     exit(1);
 	}
 	if (lseek(fd, 1024, SEEK_SET) != 1024) {
 	     perror(dev);
-	     fprintf (stderr, _("e2label: cannot seek to superblock\n"));
+	     write(2, "e2label: cannot seek to superblock\n", 35);
 	     exit(1);
 	}
 	if (read(fd, (char *) &sb, sizeof(sb)) != sizeof(sb)) {
 	     perror(dev);
-	     fprintf (stderr, _("e2label: error reading superblock\n"));
+	     write(2, "e2label: error reading superblock\n", 34);
 	     exit(1);
 	}
 	if (sb.s_magic[0] + 256*sb.s_magic[1] != EXT2_SUPER_MAGIC) {
-	     fprintf (stderr, _("e2label: not an ext2 filesystem\n"));
+	     write(2, "e2label: not an ext2 filesystem\n", 32);
 	     exit(1);
 	}
 
@@ -82,7 +84,8 @@
 	open_e2fs (dev, O_RDONLY);
 	strncpy(label, sb.s_volume_name, VOLNAMSZ);
 	label[VOLNAMSZ] = 0;
-	printf("%s\n", label);
+	write(1, label, strlen(label));
+	write(1, "\n",1);
 }
 
 static void change_label (char *dev, char *label)
@@ -93,15 +96,15 @@
 	memset(sb.s_volume_name, 0, VOLNAMSZ);
 	strncpy(sb.s_volume_name, label, VOLNAMSZ);
 	if (strlen(label) > VOLNAMSZ)
-		fprintf(stderr, _("Warning: label too long, truncating.\n"));
+		write(2, "Warning: label too long, truncating.\n", 37);
 	if (lseek(fd, 1024, SEEK_SET) != 1024) {
 	     perror(dev);
-	     fprintf (stderr, _("e2label: cannot seek to superblock again\n"));
+	     write(2, "e2label: cannot seek to superblock again\n", 41);
 	     exit(1);
 	}
 	if (write(fd, (char *) &sb, sizeof(sb)) != sizeof(sb)) {
 	     perror(dev);
-	     fprintf (stderr, _("e2label: error writing superblock\n"));
+	     write(2, "e2label: error writing superblock\n", 34);
 	     exit(1);
 	}
 }
@@ -113,7 +116,7 @@
 	else if (argc == 3)
 	     change_label(argv[1], argv[2]);
 	else {
-	     fprintf(stderr, _("Usage: e2label device [newlabel]\n"));
+	     write(2, "Usage: e2label device [newlabel]\n", 33);
 	     exit(1);
 	}
 	return 0;
