diff -up openswan-2.4.12/programs/pluto/keys.c.big-endian openswan-2.4.12/programs/pluto/keys.c
--- openswan-2.4.12/programs/pluto/keys.c.big-endian	2009-03-27 11:02:13.000000000 +0100
+++ openswan-2.4.12/programs/pluto/keys.c	2009-03-27 11:03:16.000000000 +0100
@@ -367,9 +367,9 @@ static const struct secret *
 get_secret(const struct connection *c, enum PrivateKeyKind kind, bool asym)
 {
     enum {	/* bits */
-	match_default = 01,
-	match_him = 02,
-	match_me = 04
+	match_default = 01U,
+	match_him = 02U,
+	match_me = 04U
     };
     unsigned char idstr1[IDTOA_BUF], idme[IDTOA_BUF]
 	, idhim[IDTOA_BUF], idhim2[IDTOA_BUF];
