--- linux-2.4.27/dazuko-2.0.4/configure.dazuko	Sun Oct 24 20:39:46 2004
+++ linux-2.4.27/dazuko-2.0.4/configure	Wed Jan 12 13:27:55 2005
@@ -532,11 +532,11 @@
 
 	echo "" >> Makefile
 	echo "dazuko_xp.o: dazuko_xp.c dazuko_xp.h" >> Makefile
-	echo "	\$(CC) \$(KFLAGS) -D__NO_VERSION__ -c dazuko_xp.c" >> Makefile
+	echo "	\$(CC) \$(KFLAGS) \$(ARCHFLAGS) -D__NO_VERSION__ -c dazuko_xp.c" >> Makefile
 
 	echo "" >> Makefile
 	echo "dazuko_linux.o: dazuko_linux.c dazuko_linux.h" >> Makefile
-	echo "	\$(CC) \$(KFLAGS) -c dazuko_linux.c" >> Makefile
+	echo "	\$(CC) \$(KFLAGS) \$(ARCHFLAGS) -c dazuko_linux.c" >> Makefile
 
 	echo "" >> Makefile
 	echo "dazuko.o: dazuko_xp.o dazuko_linux.o" >> Makefile
