--- linux-2.4.25/dazuko-1.2.3/configure.dazuko	Mon Jan  5 15:16:31 2004
+++ linux-2.4.25/dazuko-1.2.3/configure	Fri Sep 17 15:12:32 2004
@@ -240,7 +244,7 @@
 	echo "" >> Makefile
 
 	echo "dazuko.o: dazuko.c dazuko.h" >> Makefile
-	echo "	\$(CC) \$(KFLAGS) -c dazuko.c" >> Makefile
+	echo "	\$(CC) \$(KFLAGS) \$(ARCHFLAGS) -c dazuko.c" >> Makefile
 	echo "" >> Makefile
 
 	echo "dazukoio.o: dazukoio.c dazukoio.h" >> Makefile
