--- linux-2.4.25/net/ipsec/pfkey_v2_parser.c.compound	2004-10-14 21:19:07.263714096 +0200
+++ linux-2.4.25/net/ipsec/pfkey_v2_parser.c	2004-10-14 21:42:13.008048936 +0200
@@ -501,6 +501,7 @@
 				    s->sa_family);
 		}
 	default:
+		;
 	}
 	
 	/* XXX check if port!=0 */
@@ -4028,6 +4029,7 @@
 		}
 		break;
 	default:
+		;
 	}
 	
 	/* The NULL below causes the default extension parsers to be used */
--- linux-2.4.25/net/ipsec/ipsec_tunnel.c.compound	2004-10-14 21:19:03.755247464 +0200
+++ linux-2.4.25/net/ipsec/ipsec_tunnel.c	2004-10-14 21:19:08.850472872 +0200
@@ -1033,6 +1033,7 @@
 				stats->tx_dropped++;
 			}
 			default:
+				;
 				/* XXX what do we do with an unknown shunt spi? */
 			} /* switch (ntohl(outgoing_said.spi)) */
 			goto cleanup;
@@ -1933,6 +1934,7 @@
 					iph->tos = 0;
 					break;
 				default:
+					;
 				}
 #ifdef NET_21
 #ifdef NETDEV_23
--- linux-2.4.25/net/ipsec/libfreeswan/pfkey_v2_parse.c.compound	2004-10-14 21:19:06.376848920 +0200
+++ linux-2.4.25/net/ipsec/libfreeswan/pfkey_v2_parse.c	2004-10-14 21:19:08.877468768 +0200
@@ -1255,6 +1255,7 @@
 			SENDERR(EINVAL);
 		}
 	default:
+		;
 	}
 	
 	/* errno must not be set in downward messages */
@@ -1502,6 +1503,7 @@
 			}
 			break;
 		default:
+			;
 		}
 		if(ntohl(((struct sadb_sa*)(extensions[SADB_EXT_SA]))->sadb_sa_spi) <= 255) {
 			DEBUGGING(PF_KEY_DEBUG_PARSE_PROBLEM,
@@ -1510,7 +1512,8 @@
 				ntohl(((struct sadb_sa*)(extensions[SADB_EXT_SA]))->sadb_sa_spi));
 			SENDERR(EINVAL);
 		}
-	default:	
+	default:
+		;
 	}
 errlab:
 
--- linux-2.4.25/net/ipsec/ipsec_rcv.c.compound	2004-10-14 21:19:03.592272240 +0200
+++ linux-2.4.25/net/ipsec/ipsec_rcv.c	2004-10-14 21:19:08.902464968 +0200
@@ -1321,6 +1321,7 @@
 				break;
 #endif /* CONFIG_IPSEC_ENC_3DES */
 			default:
+				;
 			}
 			next_header = idat[ilen - 1];
 			padlen = idat[ilen - 2];
