--- linux/fritz-dsl/src.drv/makefile.dsl	Tue Jan  6 13:45:23 2004
+++ linux/fritz-dsl/src.drv/makefile	Tue Jan  6 13:48:49 2004
@@ -1,12 +1,13 @@
+KVER            = `uname -r`
 SOURCES		= main.c driver.c tables.c queue.c lib.c tools.c devif.c
 OBJECTS		= $(patsubst %.c,%.o,$(SOURCES)) 
 LIBRARY		= ../lib/$(CARD)-lib.o
 
-CARD_PATH	= /lib/modules/`uname -r`/misc
-CS_PATH		= /lib/modules/`uname -r`/pcmcia-external
+CARD_PATH	= $(BUILDROOT)/lib/modules/$(KVER)/misc
+CS_PATH		= $(BUILDROOT)/lib/modules/$(KVER)/pcmcia-external
 
 # As propsed by /usr/include/linux/version.h...
-KRNLINCL	= /lib/modules/`uname -r`/build/include
+KRNLINCL	= $(BUILDROOT)/lib/modules/$(KVER)/build/include
 # KRNLINCL	= /usr/src/linux/include
 
 DEFINES		= -DMODULE -D__KERNEL__ -DNDEBUG \
