--- openntpd-3.9p1/ntpd.h.settime	2007-01-29 19:23:56.000000000 +0100
+++ openntpd-3.9p1/ntpd.h	2007-01-29 19:24:43.000000000 +0100
@@ -56,8 +56,8 @@
 
 #define	QUERYTIME_MAX		15	/* single query might take n secs max */
 #define	OFFSET_ARRAY_SIZE	8
-#define	SETTIME_MIN_OFFSET	180	/* min offset for settime at start */
-#define	SETTIME_TIMEOUT		15	/* max seconds to wait with -s */
+#define	SETTIME_MIN_OFFSET	12 	/* min offset for settime at start */
+#define	SETTIME_TIMEOUT		30	/* max seconds to wait with -s */
 #define	LOG_NEGLIGEE		128	/* negligible drift to not log (ms) */
 
 enum client_state {
