--- ppp-2.4.5/pppd/pathnames.h.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/pathnames.h	2010-11-17 20:51:52.000000000 +0100
@@ -57,7 +57,7 @@
 
 #ifdef PLUGIN
 #ifdef __STDC__
-#define _PATH_PLUGIN	DESTDIR "/lib/pppd/" VERSION
+#define _PATH_PLUGIN	PREFIX "/lib/pppd/" VERSION
 #else /* __STDC__ */
 #define _PATH_PLUGIN	"/usr/lib/pppd"
 #endif /* __STDC__ */
--- ppp-2.4.5/pppd/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -4,10 +4,10 @@
 #
 
 # Default installation locations
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-MANDIR = $(DESTDIR)/share/man/man8
-INCDIR = $(DESTDIR)/include
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+MANDIR = $(PREFIX)/share/man/man8
+INCDIR = $(DESTDIR)@INCDIR@
 
 TARGETS = pppd
 
@@ -79,7 +79,7 @@ INCLUDE_DIRS= -I../include
 
 COMPILE_FLAGS= -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
 
-CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) '-DDESTDIR="@DESTDIR@"'
+CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) '-DPREFIX="@PREFIX@"'
 
 ifdef CHAPMS
 CFLAGS   += -DCHAPMS=1
--- ppp-2.4.5/pppd/plugins/rp-pppoe/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/plugins/rp-pppoe/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -14,9 +14,9 @@
 # $Id: Makefile.linux,v 1.8 2008/06/09 08:34:23 paulus Exp $
 #***********************************************************************
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-LIBDIR = $(DESTDIR)/lib/pppd/$(PPPDVERSION)
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+LIBDIR = $(PREFIX)/lib/pppd/$(PPPDVERSION)
 
 PPPDVERSION = $(shell awk -F '"' '/VERSION/ { print $$2; }' ../../patchlevel.h)
 
--- ppp-2.4.5/pppd/plugins/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/plugins/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -4,10 +4,10 @@ CFLAGS	= $(COPTS) -I.. -I../../include -
 LDFLAGS	= -shared
 INSTALL	= install
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-MANDIR = $(DESTDIR)/share/man/man8
-LIBDIR = $(DESTDIR)/lib/pppd/$(VERSION)
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+MANDIR = $(PREFIX)/share/man/man8
+LIBDIR = $(PREFIX)/lib/pppd/$(VERSION)
 
 SUBDIRS := rp-pppoe pppoatm pppol2tp
 # Uncomment the next line to include the radius authentication plugin
--- ppp-2.4.5/pppd/plugins/radius/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/plugins/radius/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -3,9 +3,9 @@
 # Copyright 2002 Roaring Penguin Software Inc.
 #
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-MANDIR = $(DESTDIR)/share/man/man8
-LIBDIR = $(DESTDIR)/lib/pppd/$(VERSION)
+PREFIX = $(DESTDIR)@PREFIX@
+MANDIR = $(PREFIX)/share/man/man8
+LIBDIR = $(PREFIX)/lib/pppd/$(VERSION)
 
 VERSION = $(shell awk -F '"' '/VERSION/ { print $$2; }' ../../patchlevel.h)
 
--- ppp-2.4.5/pppd/plugins/pppoatm/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/plugins/pppoatm/Makefile.linux	2010-11-17 20:57:16.000000000 +0100
@@ -6,8 +6,8 @@ INSTALL	= install
 
 #***********************************************************************
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-LIBDIR = $(DESTDIR)/lib/pppd/$(VERSION)
+PREFIX = $(DESTDIR)@PREFIX@
+LIBDIR = $(PREFIX)/lib/pppd/$(VERSION)
 
 VERSION = $(shell awk -F '"' '/VERSION/ { print $$2; }' ../../patchlevel.h)
 
--- ppp-2.4.5/pppd/Makefile.sol2.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppd/Makefile.sol2	2010-11-17 20:51:52.000000000 +0100
@@ -5,7 +5,7 @@
 
 include ../Makedefs.com
 
-CFLAGS	=  -I../include -DSVR4 -DSOL2 $(COPTS) '-DDESTDIR="@DESTDIR@"'
+CFLAGS	=  -I../include -DSVR4 -DSOL2 $(COPTS) '-DPREFIX="@PREFIX@"'
 LIBS	= -lsocket -lnsl
 
 OBJS	=  main.o magic.o fsm.o lcp.o ipcp.o upap.o chap-new.o eap.o md5.o \
--- ppp-2.4.5/chat/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/chat/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -1,8 +1,8 @@
 #	$Id: Makefile.linux,v 1.15 2006/06/04 05:07:46 paulus Exp $
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-MANDIR = $(DESTDIR)/share/man/man8
+DESTDIR = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+MANDIR = $(PREFIX)/share/man/man8
 
 CDEF1=	-DTERMIOS			# Use the termios structure
 CDEF2=	-DSIGTYPE=void			# Standard definition
--- ppp-2.4.5/pppdump/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppdump/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -1,6 +1,6 @@
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-MANDIR = $(DESTDIR)/share/man/man8
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+MANDIR = $(PREFIX)/share/man/man8
 
 CFLAGS= -O -I../include/net
 OBJS = pppdump.o bsd-comp.o deflate.o zlib.o
--- ppp-2.4.5/configure.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/configure	2010-11-17 20:51:52.000000000 +0100
@@ -2,8 +2,10 @@
 # $Id: configure,v 1.38 2008/06/15 07:08:49 paulus Exp $
 
 # Where to install stuff by default
-DESTDIR=/usr/local
+PREFIX=/usr/local
 SYSCONF=/etc
+LIBDIR=${PREFIX}/lib
+INCDIR=${PREFIX}/include
 
 #  if [ -d /NextApps ]; then
 #    system="NeXTStep"
@@ -166,7 +168,7 @@ while [ $# -gt 0 ]; do
 	    val=`expr "x$arg" : 'x[^=]*=\(.*\)'`
 	    arg=`expr "x$arg" : 'x\([^=]*\)=.*'`
 	    ;;
-	--prefix|--sysconf)
+	--prefix|--sysconf|--libdir)
 	    if [ $# -eq 0 ]; then
 		echo "error: the $arg argument requires a value" 1>&2
 		exit 1
@@ -176,16 +178,24 @@ while [ $# -gt 0 ]; do
 	    ;;
     esac
     case $arg in
-	--prefix)	DESTDIR=$val ;;
+	--prefix)	PREFIX=$val ;;
+	--libdir)	LIBDIR=$val ;;
+	--incdir)	INCDIR=$val ;;
 	--sysconfdir)	SYSCONF=$val ;;
     esac
 done
 
+
+
 mkmkf() {
     rm -f $2
     if [ -f $1 ]; then
 	echo "  $2 <= $1"
-	sed -e "s,@DESTDIR@,$DESTDIR,g" -e "s,@SYSCONF@,$SYSCONF,g" $1 >$2
+	sed -e "s,@PREFIX@,$PREFIX,g"   \
+	    -e "s,@SYSCONF@,$SYSCONF,g" \
+	    -e "s,@LIBDIR@,$LIBDIR,g"   \
+	    -e "s,@INCDIR@,$INCDIR,g" $1 >$2
+
     fi
 }
 
--- ppp-2.4.5/solaris/Makedefs.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/solaris/Makedefs	2010-11-17 20:51:52.000000000 +0100
@@ -4,8 +4,8 @@
 
 INSTALL= /usr/sbin/install
 
-BINDIR = @DESTDIR@/bin
-MANDIR = @DESTDIR@/man
+BINDIR = @PREFIX@/bin
+MANDIR = @PREFIX@/man
 ETCDIR = @SYSCONF@/ppp
 
 CC = /opt/SUNWspro/bin/cc
--- ppp-2.4.5/solaris/Makedefs.gcc.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/solaris/Makedefs.gcc	2010-11-17 20:51:52.000000000 +0100
@@ -4,8 +4,8 @@
 
 INSTALL= /usr/sbin/install
 
-BINDIR = @DESTDIR@/bin
-MANDIR = @DESTDIR@/man
+BINDIR = @PREFIX@/bin
+MANDIR = @PREFIX@/man
 ETCDIR = @SYSCONF@/ppp
 
 CC = gcc
--- ppp-2.4.5/linux/Makefile.top.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/linux/Makefile.top	2010-11-17 20:51:52.000000000 +0100
@@ -1,9 +1,9 @@
 # PPP top-level Makefile for Linux.
 
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-INCDIR = $(DESTDIR)/include
-MANDIR = $(DESTDIR)/share/man
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+INCDIR = $(DESTDIR)@INCDIR@
+MANDIR = $(PREFIX)/share/man
 ETCDIR = $(INSTROOT)@SYSCONF@/ppp
 
 # uid 0 = root
--- ppp-2.4.5/pppstats/Makefile.linux.destdir	2009-11-16 23:26:07.000000000 +0100
+++ ppp-2.4.5/pppstats/Makefile.linux	2010-11-17 20:51:52.000000000 +0100
@@ -2,9 +2,9 @@
 # pppstats makefile
 # $Id: Makefile.linux,v 1.9 2006/06/04 05:07:46 paulus Exp $
 #
-DESTDIR = $(INSTROOT)@DESTDIR@
-BINDIR = $(DESTDIR)/sbin
-MANDIR = $(DESTDIR)/share/man/man8
+PREFIX = $(DESTDIR)@PREFIX@
+BINDIR = $(PREFIX)/sbin
+MANDIR = $(PREFIX)/share/man/man8
 
 PPPSTATSRCS = pppstats.c
 PPPSTATOBJS = pppstats.o
