--- ppp-2.4.4/include/net/if_ppp.h.diet	2002-12-06 10:49:15.000000000 +0100
+++ ppp-2.4.4/include/net/if_ppp.h	2008-07-22 15:06:36.000000000 +0200
@@ -45,6 +45,15 @@
 #ifndef _IF_PPP_H_
 #define _IF_PPP_H_
 
+#include <sys/types.h>
+#include <sys/cdefs.h>
+
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <net/ppp_defs.h>
+
+__BEGIN_DECLS
+
 /*
  * Bit definitions for flags.
  */
@@ -99,12 +108,12 @@ struct ppp_option_data {
 };
 
 struct ifpppstatsreq {
-    char ifr_name[IFNAMSIZ];
+    struct ifreq	   b;
     struct ppp_stats stats;
 };
 
 struct ifpppcstatsreq {
-    char ifr_name[IFNAMSIZ];
+    struct ifreq	   b;
     struct ppp_comp_stats stats;
 };
 
@@ -153,4 +162,7 @@ struct ifpppcstatsreq {
 void pppattach __P((void));
 void pppintr __P((void));
 #endif
+
+__END_DECLS
+
 #endif /* _IF_PPP_H_ */
--- ppp-2.4.4/pppd/plugins/rp-pppoe/Makefile.linux.diet	2008-07-22 15:06:36.000000000 +0200
+++ ppp-2.4.4/pppd/plugins/rp-pppoe/Makefile.linux	2008-07-22 15:06:36.000000000 +0200
@@ -25,7 +25,7 @@ INSTALL	= install
 VERSION=3.3
 
 COPTS=-O2 -g
-CFLAGS=$(COPTS) -I../../../include/linux
+CFLAGS=$(COPTS) -I../../../include/linux -D_LINUX_IF_ETHER_H
 all: rp-pppoe.so pppoe-discovery
 
 pppoe-discovery: libplugin.a pppoe-discovery.o
@@ -47,7 +47,7 @@ clean:
 	rm -f *.o *.so
 
 plugin.o: plugin.c
-	$(CC) '-DRP_VERSION="$(VERSION)"' $(CFLAGS) -I../../..  -c -o plugin.o -fPIC plugin.c
+	$(CC) '-DRP_VERSION="$(VERSION)"' $(CFLAGS) -I../../.. -I../../../include  -c -o plugin.o -fPIC plugin.c
 
 libplugin.a: discovery.o if.o common.o debug.o
 	$(AR) -rc $@ $^
--- ppp-2.4.4/pppd/plugins/rp-pppoe/plugin.c.diet	2006-05-30 01:29:16.000000000 +0200
+++ ppp-2.4.4/pppd/plugins/rp-pppoe/plugin.c	2008-07-22 15:06:36.000000000 +0200
@@ -64,6 +64,9 @@ static char *acName = NULL;
 static char *existingSession = NULL;
 static int printACNames = 0;
 
+/* prototypes */
+void pppoe_check_options(void);
+
 static int PPPoEDevnameHook(char *cmd, char **argv, int doit);
 static option_t Options[] = {
     { "device name", o_wild, (void *) &PPPoEDevnameHook,
@@ -232,7 +235,44 @@ PPPOEDeviceOptions(void)
 
 }
 
-struct channel pppoe_channel;
+struct channel *pppoe_channel = NULL;
+
+#if 0
+former static initialization
+struct channel pppoe_channel = {
+    options: Options,
+    process_extra_options: &PPPOEDeviceOptions,
+    check_options: pppoe_check_options,
+    connect: &PPPOEConnectDevice,
+    disconnect: &PPPOEDisconnectDevice,
+    establish_ppp: &generic_establish_ppp,
+    disestablish_ppp: &generic_disestablish_ppp,
+    send_config: NULL,
+    recv_config: &PPPOERecvConfig,
+    close: NULL,
+    cleanup: NULL
+};
+#endif
+struct channel *init_channel(void)
+{
+  struct channel *ch = NULL;
+
+  if(ch = (struct channel *)malloc(sizeof(struct channel))) {
+    ch->options = Options;
+    ch->process_extra_options = PPPOEDeviceOptions;
+    ch->check_options = pppoe_check_options;
+    ch->connect = PPPOEConnectDevice;
+    ch->disconnect = PPPOEDisconnectDevice;
+    ch->establish_ppp = generic_establish_ppp;
+    ch->disestablish_ppp = generic_disestablish_ppp;
+    ch->send_config = NULL;
+    ch->recv_config = PPPOERecvConfig;
+    ch->close = NULL;
+    ch->cleanup = NULL;
+  }
+
+  return ch;
+}
 
 /**********************************************************************
  * %FUNCTION: PPPoEDevnameHook
@@ -290,9 +330,10 @@ PPPoEDevnameHook(char *cmd, char **argv,
     close(fd);
     if (r && doit) {
 	strncpy(devnam, cmd, sizeof(devnam));
-	if (the_channel != &pppoe_channel) {
+	if (!pppoe_channel) {
 
-	    the_channel = &pppoe_channel;
+            pppoe_channel = init_channel();
+	    the_channel = pppoe_channel;
 	    modem = 0;
 
 	    PPPOEInitDevice();
@@ -404,17 +445,3 @@ void pppoe_check_options(void)
     ccp_allowoptions[0].bsd_compress = 0;
     ccp_wantoptions[0].bsd_compress = 0;
 }
-
-struct channel pppoe_channel = {
-    options: Options,
-    process_extra_options: &PPPOEDeviceOptions,
-    check_options: pppoe_check_options,
-    connect: &PPPOEConnectDevice,
-    disconnect: &PPPOEDisconnectDevice,
-    establish_ppp: &generic_establish_ppp,
-    disestablish_ppp: &generic_disestablish_ppp,
-    send_config: NULL,
-    recv_config: &PPPOERecvConfig,
-    close: NULL,
-    cleanup: NULL
-};
--- ppp-2.4.4/pppd/plugins/rp-pppoe/discovery.c.diet	2005-03-22 11:22:32.000000000 +0100
+++ ppp-2.4.4/pppd/plugins/rp-pppoe/discovery.c	2008-07-22 15:06:36.000000000 +0200
@@ -598,7 +598,7 @@ discovery(PPPoEConnection *conn)
     do {
 	padiAttempts++;
 	if (padiAttempts > MAX_PADI_ATTEMPTS) {
-	    warn("Timeout waiting for PADO packets");
+	    fprintf(stderr, "Timeout waiting for PADO packets: %s\n", strerror(errno));
 	    close(conn->discoverySocket);
 	    conn->discoverySocket = -1;
 	    return;
@@ -627,7 +627,7 @@ discovery(PPPoEConnection *conn)
     do {
 	padrAttempts++;
 	if (padrAttempts > MAX_PADI_ATTEMPTS) {
-	    warn("Timeout waiting for PADS packets");
+	    fprintf(stderr, "Timeout waiting for PADO packets: %s\n", strerror(errno));
 	    close(conn->discoverySocket);
 	    conn->discoverySocket = -1;
 	    return;
--- ppp-2.4.4/pppd/Makefile.linux.diet	2008-07-22 15:06:36.000000000 +0200
+++ ppp-2.4.4/pppd/Makefile.linux	2008-07-22 15:06:36.000000000 +0200
@@ -77,7 +77,7 @@ MAXOCTETS=y
 
 INCLUDE_DIRS= -I../include
 
-COMPILE_FLAGS= -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP
+COMPILE_FLAGS= -DHAVE_PATHS_H -DHAVE_MMAP
 
 CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS) '-DPREFIX="@PREFIX@"'
 
@@ -117,10 +117,10 @@ CFLAGS   += -DHAS_SHADOW
 #LIBS     += -lshadow $(LIBS)
 endif
 
-ifneq ($(wildcard /usr/include/crypt.h),)
-CFLAGS  += -DHAVE_CRYPT_H=1
-LIBS	+= -lcrypt
-endif
+#ifneq ($(wildcard /usr/include/crypt.h),)
+#CFLAGS  += -DHAVE_CRYPT_H=1
+#LIBS	+= -lcrypt
+#endif
 
 ifdef NEEDDES
 ifndef USE_CRYPT
--- ppp-2.4.4/pppd/pppcrypt.h.diet	2002-12-04 23:44:07.000000000 +0100
+++ ppp-2.4.4/pppd/pppcrypt.h	2008-07-22 15:06:36.000000000 +0200
@@ -35,6 +35,8 @@
 
 #ifdef HAVE_CRYPT_H
 #include <crypt.h>
+#else
+#include <unistd.h>
 #endif
 
 #ifndef USE_CRYPT
--- ppp-2.4.4/pppd/sys-linux.c.diet	2005-08-27 00:44:35.000000000 +0200
+++ ppp-2.4.4/pppd/sys-linux.c	2008-07-22 15:08:31.000000000 +0200
@@ -69,11 +69,13 @@
  * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  */
 
+#define __KERNEL_STRICT_NAMES
+#define _LINUX_SOCKET_H
 #include <sys/ioctl.h>
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <sys/time.h>
-#include <sys/errno.h>
+#include <errno.h>
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/utsname.h>
@@ -102,7 +104,7 @@
 #define MAX_ADDR_LEN 7
 #endif
 
-#if __GLIBC__ >= 2
+#if __GLIBC__ >= 2 || defined(__dietlibc__)
 #include <asm/types.h>		/* glibc 2 conflicts with linux/types.h */
 #include <net/if.h>
 #include <net/if_arp.h>
@@ -453,6 +455,13 @@ int generic_establish_ppp (int fd)
     if (new_style_driver) {
 	int flags;
 
+	/* if a ppp_fd is already open, close it first */
+	if(ppp_fd > 0) {
+	    close(ppp_fd); 
+	    remove_fd(ppp_fd); 
+	    ppp_fd = -1; 
+	}
+
 	/* Open an instance of /dev/ppp and connect the channel to it */
 	if (ioctl(fd, PPPIOCGCHAN, &chindex) == -1) {
 	    error("Couldn't get channel number: %m");
--- ppp-2.4.4/pppd/main.c.diet	2006-06-04 05:52:50.000000000 +0200
+++ ppp-2.4.4/pppd/main.c	2008-07-22 15:06:36.000000000 +0200
@@ -88,6 +88,7 @@
 #include <sys/resource.h>
 #include <sys/stat.h>
 #include <sys/socket.h>
+#include <sys/utsname.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
 
@@ -314,6 +315,7 @@ main(argc, argv)
     struct passwd *pw;
     struct protent *protp;
     char numbuf[16];
+    struct utsname u;
 
     link_stats_valid = 0;
     new_phase(PHASE_INITIALIZE);
@@ -323,11 +325,13 @@ main(argc, argv)
     /* Initialize syslog facilities */
     reopen_log();
 
-    if (gethostname(hostname, MAXNAMELEN) < 0 ) {
+    uname(&u);
+    if (!u.nodename) {
 	option_error("Couldn't get hostname: %m");
 	exit(1);
     }
-    hostname[MAXNAMELEN-1] = 0;
+    strncpy(hostname, u.nodename, MAXNAMELEN);
+    hostname[MAXNAMELEN-1] = (char)'\0';
 
     /* make sure we don't create world or group writable files. */
     umask(umask(0777) | 022);
