diff -up ucspi-tcp-0.88/Makefile.cross ucspi-tcp-0.88/Makefile
--- ucspi-tcp-0.88/Makefile.cross	2008-03-18 17:55:25.000000000 +0100
+++ ucspi-tcp-0.88/Makefile	2008-03-18 17:56:43.000000000 +0100
@@ -149,14 +149,6 @@ check: \
 it instcheck
 	./instcheck
 
-chkshsgr: \
-load chkshsgr.o
-	./load chkshsgr 
-
-chkshsgr.o: \
-compile chkshsgr.c exit.h
-	./compile chkshsgr.c
-
 choose: \
 warn-auto.sh choose.sh conf-home
 	cat warn-auto.sh choose.sh \
@@ -345,9 +337,7 @@ choose compile load trysgprm.c hassgprm.
 	./choose cl trysgprm hassgprm.h1 hassgprm.h2 > hassgprm.h
 
 hasshsgr.h: \
-choose compile load tryshsgr.c hasshsgr.h1 hasshsgr.h2 chkshsgr \
-warn-shsgr
-	./chkshsgr || ( cat warn-shsgr; exit 1 )
+choose compile load tryshsgr.c hasshsgr.h1 hasshsgr.h2 
 	./choose clr tryshsgr hasshsgr.h1 hasshsgr.h2 > hasshsgr.h
 
 haswaitp.h: \
