--- dietlibc-0.32/libdl/_dl_load.c.gcc4	2002-08-30 14:25:06.000000000 +0200
+++ dietlibc-0.32/libdl/_dl_load.c	2009-06-10 21:22:33.000000000 +0200
@@ -4,6 +4,7 @@
 #include <sys/stat.h>
 #include <dlfcn.h>
 #include <elf.h>
+#include <alloca.h>
 
 #include "_dl_int.h"
 
@@ -334,10 +335,11 @@ struct _dl_handle* _dl_dyn_scan(struct _
   /* do PTL / GOT relocation */
   if (pltreltype == DT_REL) {
     Elf_Rel *tmp = jmprel;
+    char * t=(char*)tmp;
 #ifdef DEBUG
     pf(__func__": rel got\n");
 #endif
-    for (;(char*)tmp<(((char*)jmprel)+pltrelsize);(char*)tmp=((char*)tmp)+sizeof(Elf_Rel)) {
+    for (;t < (((char*)jmprel)+pltrelsize); t = t + sizeof(Elf_Rel)) {
       if ((flags&RTLD_NOW)) {
 	unsigned long sym=(unsigned long)_dl_sym(dh,ELF_R_SYM(tmp->r_info));
 	if (sym) *((unsigned long*)(dh->mem_base+tmp->r_offset))=sym;
@@ -357,10 +359,11 @@ struct _dl_handle* _dl_dyn_scan(struct _
   }
   if (pltreltype == DT_RELA) {
     Elf_Rela *tmp = jmprel;
+    char * t=(char*)tmp;
 #ifdef DEBUG
     pf(__func__": rela got\n");
 #endif
-    for (;(char*)tmp<(((char*)jmprel)+pltrelsize);(char*)tmp=((char*)tmp)+sizeof(Elf_Rela)) {
+    for (;t < (((char*)jmprel)+pltrelsize); t = t + sizeof(Elf_Rela)) {
       if ((flags&RTLD_NOW)) {
 	unsigned long sym=(unsigned long)_dl_sym(dh,ELF_R_SYM(tmp->r_info));
 	if (sym) *((unsigned long*)(dh->mem_base+tmp->r_offset))=sym;
--- dietlibc-0.32/libdl/_dl_int.h.gcc4	2007-11-09 20:13:14.000000000 +0100
+++ dietlibc-0.32/libdl/_dl_int.h	2009-06-10 21:22:33.000000000 +0200
@@ -114,12 +114,10 @@ struct r_debug {
 #define GNU_HASH_CHAIN(p,n)	((p)+(1+GNU_HASH_BUCKET_LEN(p)+(n)))
 
 /* _dl_alloc.c */
-#if 0
+#ifndef __DIET_LD_SO__
 extern struct _dl_handle* _dl_root_handle;
 extern struct _dl_handle* _dl_top_handle;
 extern struct _dl_handle* _dl_free_list;
-#endif
-#ifndef __DIET_LD_SO__
 void _dl_free_handle(struct _dl_handle* dh);
 struct _dl_handle* _dl_get_handle();
 struct _dl_handle* _dl_find_lib(const char* name);
@@ -137,8 +135,10 @@ const char* _dl_get_rpath();
 int _dl_search(char *buf, int len, const char *filename);
 
 /* dlsym.c */
+#ifdef __DIET_LD_SO__
 void *_dlsym(void*dh,const char*symbol);
 void *_dl_sym_search_str(struct _dl_handle*h,const char*name);
+#endif
 void *_dl_sym(struct _dl_handle * h, int symbol);
 void *_dl_sym_next(struct _dl_handle * h, int symbol);
 
