--- dietlibc-stage1/libugly/gai_strerror.c.netdb	2007-06-06 20:40:05.000000000 +0200
+++ dietlibc-stage1/libugly/gai_strerror.c	2007-06-06 20:40:31.000000000 +0200
@@ -1,4 +1,4 @@
-#include <sys/socket.h>
+#include <netdb.h>
 
 const char* gai_strerror(int error) {
   switch (error) {
--- dietlibc-stage1/libugly/freeaddrinfo.c.netdb	2001-05-18 21:15:01.000000000 +0200
+++ dietlibc-stage1/libugly/freeaddrinfo.c	2007-06-06 20:38:19.000000000 +0200
@@ -1,4 +1,4 @@
-#include <sys/socket.h>
+#include <netdb.h>
 #include <stdlib.h>
 
 void freeaddrinfo(struct addrinfo *res) {
--- dietlibc-stage1/include/sys/socket.h.netdb	2005-03-15 09:51:22.000000000 +0100
+++ dietlibc-stage1/include/sys/socket.h	2007-06-06 20:38:15.000000000 +0200
@@ -401,48 +401,6 @@
 
 int socketpair(int d, int type, int protocol, int sv[2]);
 
-/* currently not supported: */
-#define NI_NOFQDN 1
-
-#define NI_NUMERICHOST 2
-#define NI_NAMEREQD 4
-#define NI_NUMERICSERV 8
-#define NI_DGRAM 16
-
-struct addrinfo {
-  int     ai_flags;
-  int     ai_family;
-  int     ai_socktype;
-  int     ai_protocol;
-  size_t  ai_addrlen;
-  struct sockaddr *ai_addr;
-  char   *ai_canonname;
-  struct addrinfo *ai_next;
-};
-
-int getnameinfo(const struct sockaddr *sa, socklen_t salen, char *host,
-		size_t hostlen, char *serv, size_t servlen, int flags) __THROW;
-int getaddrinfo(const char *node, const char *service, const struct
-		addrinfo *hints, struct addrinfo **res) __THROW;
-void freeaddrinfo(struct addrinfo *res) __THROW;
-const char *gai_strerror(int errcode) __THROW;
-
-#define EAI_FAMILY -1
-#define EAI_SOCKTYPE -2
-#define EAI_BADFLAGS -3
-#define EAI_NONAME -4
-#define EAI_SERVICE -5
-#define EAI_ADDRFAMILY -6
-#define EAI_NODATA -7
-#define EAI_MEMORY -8
-#define EAI_FAIL -9
-#define EAI_AGAIN -10
-#define EAI_SYSTEM -11
-
-#define AI_NUMERICHOST 1
-#define AI_CANONNAME 2
-#define AI_PASSIVE 4
-
 /* Linux-specific socket ioctls */
 #define SIOCINQ		FIONREAD
 #define SIOCOUTQ	TIOCOUTQ
--- dietlibc-stage1/include/netdb.h.netdb	2003-08-19 18:58:17.000000000 +0200
+++ dietlibc-stage1/include/netdb.h	2007-06-06 20:38:14.000000000 +0200
@@ -123,6 +123,55 @@
 #define NI_MAXHOST 1025
 #define NI_MAXSERV 32
 
+/* Possible values for `ai_flags' field in `addrinfo' structure. */
+# define AI_NUMERICHOST 0x0001  /* Don't use name resolution. */
+# define AI_CANONNAME   0x0002  /* Request for canonical name. */
+# define AI_PASSIVE     0x0004  /* Socket address is intended for `bind'. */
+# define AI_NUMERICSERV 0x0008  /* Inhibit service name resolution. */
+# define AI_V4MAPPED	0x0010	/* IPv4 mapped addresses are acceptable. */
+# define AI_ALL		0x0020	/* Query for both IPv4 and IPv6 addresses. */
+# define AI_ADDRCONFIG	0x0040	/* Query for IPv4 addresses only when an */
+				/* IPv4 address is configured; query for */
+				/* IPv6 addresses only when an IPv6 address */
+                                /* is configured. */
+
+/* currently not supported: */
+#define NI_NOFQDN 1
+#define NI_NUMERICHOST 2
+#define NI_NAMEREQD 4
+#define NI_NUMERICSERV 8
+#define NI_DGRAM 16
+
+struct addrinfo {
+  int     ai_flags;
+  int     ai_family;
+  int     ai_socktype;
+  int     ai_protocol;
+  size_t  ai_addrlen;
+  struct sockaddr *ai_addr;
+  char   *ai_canonname;
+  struct addrinfo *ai_next;
+};
+
+int getnameinfo(const struct sockaddr *sa, socklen_t salen, char *host,
+		size_t hostlen, char *serv, size_t servlen, int flags) __THROW;
+int getaddrinfo(const char *node, const char *service, const struct
+		addrinfo *hints, struct addrinfo **res) __THROW;
+void freeaddrinfo(struct addrinfo *res) __THROW;
+const char *gai_strerror(int errcode) __THROW;
+
+#define EAI_FAMILY -1
+#define EAI_SOCKTYPE -2
+#define EAI_BADFLAGS -3
+#define EAI_NONAME -4
+#define EAI_SERVICE -5
+#define EAI_ADDRFAMILY -6
+#define EAI_NODATA -7
+#define EAI_MEMORY -8
+#define EAI_FAIL -9
+#define EAI_AGAIN -10
+#define EAI_SYSTEM -11
+
 __END_DECLS
 
 #endif
