--- dietlibc-0.27/libshell/fnmatch.c.fnmatch	2004-10-06 13:05:50.987513960 +0200
+++ dietlibc-0.27/libshell/fnmatch.c	2004-10-06 13:06:33.307080408 +0200
@@ -69,7 +69,7 @@
 
       ++pattern;
       if (*string=='/' && flags&FNM_PATHNAME) return FNM_NOMATCH;
-      if (*pattern=='!') { neg=1; ++pattern; }
+      if (*pattern=='!' || *pattern=='^') { neg=1; ++pattern; }
       start=pattern;
       while (*pattern) {
 	int res=0;
