--- dietlibc-0.32/libcruft/dn_skipname.c.dn_skipname	2007-02-08 15:40:20.000000000 +0100
+++ dietlibc-0.32/libcruft/dn_skipname.c	2009-06-10 21:52:59.000000000 +0200
@@ -1,4 +1,33 @@
 #include <resolv.h>
+#include <errno.h>
+
+int dn_skipname(unsigned char *comp_dn, const unsigned char *eom) {
+  const unsigned char *cp = comp_dn;
+  unsigned int n;
+
+  while (cp < eom && (n = *cp) != 0) {
+    cp++;
+    switch (n & 0xc0) {
+      case 0:
+	cp += n;
+	if (*cp)
+	  continue;
+      case 0xc0:
+	cp++;
+	goto out;
+      default:
+	goto err;
+    }
+  }
+out:
+  if (cp > eom) {
+err:
+    errno = EMSGSIZE;
+    return -1;
+  }
+  return (cp - comp_dn);
+}
+#include <resolv.h>
 
 int dn_skipname(const unsigned char* cur,const unsigned char* eom) {
   const unsigned char* orig=cur;
--- dietlibc-0.32/include/resolv.h.dn_skipname	2008-10-26 19:58:17.000000000 +0100
+++ dietlibc-0.32/include/resolv.h	2009-06-10 21:52:59.000000000 +0200
@@ -123,6 +123,8 @@ int dn_expand(const unsigned char *msg, 
       const unsigned char *comp_dn, unsigned char *exp_dn,
       int length) __THROW;
 
+int dn_skipname(unsigned char *comp_dn, const unsigned char *eom) __THROW;
+
 void res_close(void) __THROW __attribute_dontuse__;
 
 int dn_skipname(const unsigned char* cur,const unsigned char* eom) __THROW;
