--- httpauth-0.5.1/daemon/httpauthd.c.mutex	2006-06-01 13:47:30.000000000 +0200
+++ httpauth-0.5.1/daemon/httpauthd.c	2006-06-01 13:47:58.000000000 +0200
@@ -188,7 +188,7 @@
 
     /* Create the main mutex */  
     if(pthread_mutexattr_init(&g_mutexattr) != 0 ||
-       pthread_mutexattr_settype(&g_mutexattr, HA_MUTEX_TYPE) ||
+       pthread_mutexattr_setkind_np(&g_mutexattr, HA_MUTEX_TYPE) ||
        pthread_mutex_init(&g_mutex, &g_mutexattr) != 0)
         errx(1, "threading problem. can't create mutex");
   
--- httpauth-0.5.1/daemon/ntlm.c.mutex	2006-06-01 13:48:11.000000000 +0200
+++ httpauth-0.5.1/daemon/ntlm.c	2006-06-01 13:48:36.000000000 +0200
@@ -631,7 +631,7 @@
     {
         /* Create the smblib mutex */  
         if(pthread_mutexattr_init(&g_smblib_mutexattr) != 0 ||
-           pthread_mutexattr_settype(&g_smblib_mutexattr, HA_MUTEX_TYPE) ||
+           pthread_mutexattr_setkind_np(&g_smblib_mutexattr, HA_MUTEX_TYPE) ||
            pthread_mutex_init(&g_smblib_mutex, &g_smblib_mutexattr) != 0)
         {
             ha_messagex(NULL, LOG_CRIT, "threading problem. can't create mutex");
--- httpauth-0.5.1/daemon/mysql.c.mutex	2006-06-01 13:50:00.000000000 +0200
+++ httpauth-0.5.1/daemon/mysql.c	2006-06-01 13:50:20.000000000 +0200
@@ -758,7 +758,7 @@
     {
         /* Create the smblib mutex */  
         if(pthread_mutexattr_init(&g_mysql_mutexattr) != 0 ||
-           pthread_mutexattr_settype(&g_mysql_mutexattr, HA_MUTEX_TYPE) ||
+           pthread_mutexattr_setkind_np(&g_mysql_mutexattr, HA_MUTEX_TYPE) ||
            pthread_mutex_init(&g_mysql_mutex, &g_mysql_mutexattr) != 0)
         {
             ha_messagex(NULL, LOG_CRIT, "threading problem. can't create mutex");
