--- qmail-1.03/auto-int8.c.builtin	2004-05-06 22:09:34.470795712 +0200
+++ qmail-1.03/auto-int8.c	2004-05-06 22:09:58.514140568 +0200
@@ -7,7 +7,7 @@
 char buf1[256];
 substdio ss1 = SUBSTDIO_FDBUF(write,1,buf1,sizeof(buf1));
 
-void puts(s)
+static void puts(s)
 char *s;
 {
   if (substdio_puts(&ss1,s) == -1) _exit(111);
--- qmail-1.03/qmail-pop3d.c.builtin	2004-05-06 22:12:19.797662208 +0200
+++ qmail-1.03/qmail-pop3d.c	2004-05-06 22:13:27.462375608 +0200
@@ -45,7 +45,7 @@
 {
   substdio_put(&ssout,buf,len);
 }
-void puts(s) char *s;
+static void puts(s) char *s;
 {
   substdio_puts(&ssout,s);
 }
--- qmail-1.03/qmail-popup.c.builtin	2004-05-06 22:13:41.884183160 +0200
+++ qmail-1.03/qmail-popup.c	2004-05-06 22:13:55.379131616 +0200
@@ -38,7 +38,7 @@
 char ssinbuf[128];
 substdio ssin = SUBSTDIO_FDBUF(saferead,0,ssinbuf,sizeof ssinbuf);
 
-void puts(s) char *s;
+static void puts(s) char *s;
 {
   substdio_puts(&ssout,s);
 }
--- qmail-1.03/qmail-inject.c.builtin	2004-05-06 22:11:44.903966856 +0200
+++ qmail-1.03/qmail-inject.c	2004-05-06 22:12:08.934313688 +0200
@@ -55,7 +55,7 @@
 
 void put(s,len) char *s; int len;
 { if (flagqueue) qmail_put(&qqt,s,len); else substdio_put(subfdout,s,len); }
-void puts(s) char *s; { put(s,str_len(s)); }
+static void puts(s) char *s; { put(s,str_len(s)); }
 
 void perm() { _exit(100); }
 void temp() { _exit(111); }
--- qmail-1.03/auto-str.c.builtin	2004-05-06 22:10:38.892002200 +0200
+++ qmail-1.03/auto-str.c	2004-05-06 22:10:51.840033800 +0200
@@ -5,7 +5,7 @@
 char buf1[256];
 substdio ss1 = SUBSTDIO_FDBUF(write,1,buf1,sizeof(buf1));
 
-void puts(s)
+static void puts(s)
 char *s;
 {
   if (substdio_puts(&ss1,s) == -1) _exit(111);
--- qmail-1.03/auto-int.c.builtin	2004-05-06 22:10:06.916863160 +0200
+++ qmail-1.03/auto-int.c	2004-05-06 22:10:21.538640312 +0200
@@ -7,7 +7,7 @@
 char buf1[256];
 substdio ss1 = SUBSTDIO_FDBUF(write,1,buf1,sizeof(buf1));
 
-void puts(s)
+static void puts(s)
 char *s;
 {
   if (substdio_puts(&ss1,s) == -1) _exit(111);
