--- linux-2.6.17.7/Documentation/networking/ifenslave.c.diet	2006-07-25 05:36:01.000000000 +0200
+++ linux-2.6.17.7/Documentation/networking/ifenslave.c	2006-07-27 10:24:46.000000000 +0200
@@ -165,9 +165,9 @@
 #include <linux/sockios.h>
 
 typedef unsigned long long u64;	/* hack, so we may include kernel's ethtool.h */
-typedef __uint32_t u32;		/* ditto */
-typedef __uint16_t u16;		/* ditto */
-typedef __uint8_t u8;		/* ditto */
+typedef uint32_t u32;		/* ditto */
+typedef uint16_t u16;		/* ditto */
+typedef uint8_t u8;		/* ditto */
 #include <linux/ethtool.h>
 
 struct option longopts[] = {
