--- checkpassword-0.90/Makefile.cross	2008-03-20 21:14:57.000000000 +0100
+++ checkpassword-0.90/Makefile	2008-03-20 21:15:30.000000000 +0100
@@ -80,14 +80,6 @@
 hasuserpw.h
 	./compile checkpassword.c
 
-chkshsgr: \
-load chkshsgr.o
-	./load chkshsgr 
-
-chkshsgr.o: \
-compile chkshsgr.c exit.h
-	./compile chkshsgr.c
-
 choose: \
 warn-auto.sh choose.sh conf-home
 	cat warn-auto.sh choose.sh \
@@ -120,9 +112,7 @@
 	./compile error_str.c
 
 hasshsgr.h: \
-choose compile load tryshsgr.c hasshsgr.h1 hasshsgr.h2 chkshsgr \
-warn-shsgr
-	./chkshsgr || ( cat warn-shsgr; exit 1 )
+choose compile load tryshsgr.c hasshsgr.h1 hasshsgr.h2
 	./choose clr tryshsgr hasshsgr.h1 hasshsgr.h2 > hasshsgr.h
 
 hasspnam.h: \
