--- rrdtool-1.4.3/configure.ac.silent_rules	2010-10-09 20:28:34.000000000 +0200
+++ rrdtool-1.4.3/configure.ac	2010-10-09 20:29:23.000000000 +0200
@@ -37,6 +37,7 @@ AC_SUBST(LIBVERS)
 AC_CANONICAL_TARGET
 m4_version_prereq(2.60, [AC_USE_SYSTEM_EXTENSIONS], [#])
 AM_INIT_AUTOMAKE
+AM_SILENT_RULES([yes])
 AM_MAINTAINER_MODE
 # Enable silent build rules by default, requires at least
 # Automake-1.11. Disable by either passing --disable-silent-rules to
