--- polarssl-0.14.0/Makefile.config	2010-06-18 18:42:26.000000000 +0200
+++ polarssl-0.14.0/Makefile	2010-09-28 21:05:22.000000000 +0200
@@ -6,8 +6,8 @@ PREFIX=polarssl_
 
 all:
 	cd library  && $(MAKE) all && cd ..
-	cd programs && $(MAKE) all && cd ..
-	cd tests && $(MAKE) all && cd ..
+	# cd programs && $(MAKE) all && cd ..
+	# cd tests && $(MAKE) all && cd ..
 
 install:
 	mkdir -p $(DESTDIR)/include/polarssl
@@ -16,14 +16,14 @@ install:
 	mkdir -p $(DESTDIR)/lib
 	cp library/libpolarssl.* $(DESTDIR)/lib
 	
-	mkdir -p $(DESTDIR)/bin
-	for p in programs/*/* ; do              \
-	    if [ -x $$p ] && [ ! -d $$p ] ;     \
-	    then                                \
-	        f=$(PREFIX)`basename $$p` ;     \
-	        cp $$p $(DESTDIR)/bin/$$f ;     \
-	    fi                                  \
-	done
+	#mkdir -p $(DESTDIR)/bin
+	#for p in programs/*/* ; do              \
+	#    if [ -x $$p ] && [ ! -d $$p ] ;     \
+	#    then                                \
+	#        f=$(PREFIX)`basename $$p` ;     \
+	#        cp $$p $(DESTDIR)/bin/$$f ;     \
+	#    fi                                  \
+	#done
 
 clean:
 	cd library  && $(MAKE) clean && cd ..
--- polarssl-0.14.0/include/polarssl/config.h.config	2010-08-16 15:09:22.000000000 +0200
+++ polarssl-0.14.0/include/polarssl/config.h	2010-09-28 21:01:44.000000000 +0200
@@ -72,13 +72,13 @@
 
 /*
  * Enable all SSL/TLS debugging messages.
- */
 #define POLARSSL_DEBUG_MSG
+ */
 
 /*
  * Enable the checkup functions (*_self_test).
- */
 #define POLARSSL_SELF_TEST
+ */
 
 /*
  * Enable run-time version information functions
@@ -144,16 +144,16 @@
  *      SSL_RSA_CAMELLIA_128_SHA
  *      SSL_RSA_CAMELLIA_256_SHA
  *      SSL_EDH_RSA_CAMELLIA_256_SHA
- */
 #define POLARSSL_CAMELLIA_C
+ */
 
 /*
  * Module:  library/certs.c
  * Caller:
  *
  * This module is used for testing (ssl_client/server).
- */
 #define POLARSSL_CERTS_C
+ */
 
 /*
  * Module:  library/debug.c
@@ -162,8 +162,8 @@
  *          library/ssl_tls.c
  *
  * This module provides debugging functions.
- */
 #define POLARSSL_DEBUG_C
+ */
 
 /*
  * Module:  library/des.c
@@ -235,8 +235,8 @@
  * Caller:  library/aes.c
  *
  * This modules adds support for the VIA PadLock on x86.
- */
 #define POLARSSL_PADLOCK_C
+ */
 
 /*
  * Module:  library/rsa.c
@@ -265,16 +265,16 @@
  * Caller:
  *
  * This module adds support for SHA-224 and SHA-256.
- */
 #define POLARSSL_SHA2_C
+ */
 
 /*
  * Module:  library/sha4.c
  * Caller:
  *
  * This module adds support for SHA-384 and SHA-512.
- */
 #define POLARSSL_SHA4_C
+ */
 
 /*
  * Module:  library/ssl_cli.c
@@ -324,13 +324,13 @@
  * Caller:
  *
  * This module is required for X.509 certificate writing.
- */
 #define POLARSSL_X509_WRITE_C
+ */
 
 /*
  * Module:  library/xtea.c
  * Caller:
- */
 #define POLARSSL_XTEA_C
+ */
 
 #endif /* config.h */
