--- keepalived-1.1.11/keepalived/vrrp/vrrp_if.c.types	2006-09-19 14:01:41.000000000 +0200
+++ keepalived-1.1.11/keepalived/vrrp/vrrp_if.c	2006-09-19 14:01:54.000000000 +0200
@@ -26,10 +26,10 @@
 #include <unistd.h>
 #include <string.h>
 #include <sys/types.h>
-typedef __uint64_t u64;
-typedef __uint32_t u32;
-typedef __uint16_t u16;
-typedef __uint8_t u8;
+typedef uint64_t u64;
+typedef uint32_t u32;
+typedef uint16_t u16;
+typedef uint8_t u8;
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <netinet/ip.h>
