--- mini_httpd-1.19/mini_httpd.c.stderr	2005-10-05 19:40:27.000000000 +0200
+++ mini_httpd-1.19/mini_httpd.c	2005-10-05 19:43:21.000000000 +0200
@@ -1836,18 +1836,25 @@
 	(void) close( p[0] );
 	if ( p[1] != STDOUT_FILENO )
 	    (void) dup2( p[1], STDOUT_FILENO );
+#ifdef EAT_STDERR
 	if ( p[1] != STDERR_FILENO )
 	    (void) dup2( p[1], STDERR_FILENO );
 	if ( p[1] != STDOUT_FILENO && p[1] != STDERR_FILENO )
 	    (void) close( p[1] );
+#else
+	if ( p[1] != STDOUT_FILENO )
+	    (void) close( p[1] );
+#endif
 	}
     else
 	{
 	/* Otherwise, the request socket is stdout/stderr. */
 	if ( conn_fd != STDOUT_FILENO )
 	    (void) dup2( conn_fd, STDOUT_FILENO );
+#ifdef EAT_STDERR
 	if ( conn_fd != STDERR_FILENO )
 	    (void) dup2( conn_fd, STDERR_FILENO );
+#endif
 	}
 
     /* At this point we would like to set conn_fd to be close-on-exec.
