--- binutils-2.17/ld/Makefile.in.ldmake	2006-06-03 06:45:50.000000000 +0200
+++ binutils-2.17/ld/Makefile.in	2006-10-15 22:54:22.000000000 +0200
@@ -269,7 +269,7 @@
 # We put the scripts in the directory $(scriptdir)/ldscripts.
 # We can't put the scripts in $(datadir) because the SEARCH_DIR
 # directives need to be different for native and cross linkers.
-scriptdir = $(tooldir)/lib
+scriptdir = $(libdir)
 BASEDIR = $(srcdir)/..
 BFDDIR = $(BASEDIR)/bfd
 INCDIR = $(BASEDIR)/include
