--- qdbm-1.8.5/Makefile.in.ldrun	2004-03-24 10:09:19.066916672 +0100
+++ qdbm-1.8.5/Makefile.in	2004-03-24 10:10:16.882127424 +0100
@@ -49,8 +49,8 @@
 LD = @LD@
 LIBS = -lqdbm @LIBS@
 LIBLDFLAGS = @LIBS@
-LDFLAGS = -L. -L/usr/local/lib $(LIBS)
-LDENV = LD_RUN_PATH=/lib:/usr/lib:$(MYLIBDIR)
+LDFLAGS = -L. $(LIBS)
+LDENV = LD_RUN_PATH=$(MYLIBDIR)
 AR = @AR@
 ARFLAGS = rcsv
 RUNENV = LD_LIBRARY_PATH=.
