--- matrixssl-1-2-5/src/os/linux/linux.c.pthread	2005-04-04 23:23:32.000000000 +0200
+++ matrixssl-1-2-5/src/os/linux/linux.c	2005-07-01 19:22:45.000000000 +0200
@@ -120,7 +120,11 @@
 */
 #ifdef USE_MULTITHREADING
 #ifndef OSX
+#ifdef __dietlibc__
+	pthread_mutexattr_setkind_np(&attr, PTHREAD_MUTEX_RECURSIVE_NP);
+#else
 	pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE_NP);
+#endif
 #endif /* !OSX */
 	pthread_mutexattr_init(&attr);
 #endif /* USE_MULTITHREADING */
--- matrixssl-1-2-5/src/os/osLayer.h.pthread	2005-07-01 19:27:23.000000000 +0200
+++ matrixssl-1-2-5/src/os/osLayer.h	2005-07-01 19:27:58.000000000 +0200
@@ -30,6 +30,8 @@
  */
 /******************************************************************************/
 
+#undef __i386__
+
 #ifndef _h_OS_LAYER
 #define _h_OS_LAYER
 
